mpi4jax/_src/tokenizer.py

```diff
@@ -0,0 +1,186 @@
+import jax
+from jax import linear_util as lu
+from jax.interpreters import xla
+
+
+token_override_registry = {}  # Dict[Callable, Callable]
+recursive_token_forwarding_registry = {}  # Dict[Primitive, Callable]
+
+
+def safe_map(f, *args):
+    args = [list(arg) for arg in args]
+    for arg in args:
+        assert len(arg) == len(args[0]), f"Mismatched lengths: {[len(x) for x in args]}"
+    return [f(*arg) for arg in zip(*args)]
+
+
+def xla_call_overrride(read, eqn, token):
+    subfuns, bind_params = eqn.primitive.get_bind_params(eqn.params)
+    bind_params["donated_invars"] = (False,) + bind_params["donated_invars"]
+    map_token = lambda func: lu.wrap_init(  # noqa: E731
+        lambda token, *args: _token_forwarding(func.call_wrapped, token)(*args)
+    )
+    subfuns = safe_map(map_token, subfuns)
+    ans = eqn.primitive.bind(
+        *subfuns, token, *safe_map(read, eqn.invars), **bind_params
+    )
+    token = ans[0]
+    ans = ans[1:]  # Drop the token.
+    return token, ans
+
+
+recursive_token_forwarding_registry[xla.xla_call_p] = xla_call_overrride
+
+
+def scan_override(read, eqn, token):
+    _, bind_params = eqn.primitive.get_bind_params(eqn.params)
+    new_body_fn = lambda token, *args: _token_forwarding(  # noqa: E731
+        jax.core.jaxpr_as_fun(bind_params["jaxpr"]), token
+    )(*args)
+    bind_params["jaxpr"] = jax.make_jaxpr(new_body_fn)(
+        token, *safe_map(read, eqn.invars)
+    )
+    # Update bind_params to account for the additional token.
+    bind_params["num_carry"] += 1
+    bind_params["linear"] = (False,) + bind_params["linear"]
+    ans = eqn.primitive.bind(token, *safe_map(read, eqn.invars), **bind_params)
+    token = ans[0]
+    ans = ans[1:]  # Drop the token.
+    return token, ans
+
+
+recursive_token_forwarding_registry[jax.lax.scan_p] = scan_override
+
+
+def while_override(read, eqn, token):
+    _, bind_params = eqn.primitive.get_bind_params(eqn.params)
+    new_body_fn = lambda token, *args: _token_forwarding(  # noqa: E731
+        jax.core.jaxpr_as_fun(bind_params["body_jaxpr"]), token
+    )(*args)
+    bind_params["body_jaxpr"] = jax.make_jaxpr(new_body_fn)(
+        token, *safe_map(read, eqn.invars)
+    )
+    # We use `auto_tokenize` here since the condition function
+    # is forced to only return a boolean value and cannot return
+    # a token.
+    new_cond_fn = lambda token, *args: auto_tokenize(  # noqa: E731
+        jax.core.jaxpr_as_fun(bind_params["cond_jaxpr"]), token
+    )(*args)
```

Comment:
That's a problem though, isn't it? The return token from the condition function is silently dropped, so we can't guarantee proper threading here?

Reply:
BTW,

```suggestion
    new_cond_fn = lambda token, *args: _token_forwarding(  # noqa: E731
        jax.core.jaxpr_as_fun(bind_params["cond_jaxpr"]), token
    )(*args)[1:]
```

and we can remove the `token` argument for `auto_vectorize`.

Reply:
`cond` is a really tough transform to make work regardless. I think XLA will default to executing *both* branches always, and then just returning the value from the correct branch. Not sure if that is still the case, but it was when I was working at Google.

Reply:
The parent comments actually relates to `while` (which also has a "condition function"). If you think there's no way to handle `cond` consistently we should just drop it from the registry, I prefer a loud exception over deadlocks and silently wrong results, and it's not that useful anyway (just a glorified `jnp.where`?).

Reply:
Oh sorry I got confused for a second.